packages/api/src/moderation/index.ts

```diff
@@ -0,0 +1,343 @@
+import {
+  takeHighestPriorityDecision,
+  downgradeDecision,
+  isModerationDecisionNoop,
+  isQuotedPost,
+  isQuotedPostWithMedia,
+  toModerationUI,
+} from './util'
+
+// profiles
+// =
+
+export interface ProfileModeration {
+  decisions: {
+    account: ModerationDecision
+    profile: ModerationDecision
+  }
+  account: ModerationUI
+  profile: ModerationUI
+  avatar: ModerationUI
+}
+
+export function moderateProfile(
+  subject: ModerationSubjectProfile,
+  opts: ModerationOpts,
+): ProfileModeration {
+  // decide the moderation the account and the profile
+  const account = decideAccount(subject, opts)
+  const profile = decideProfile(subject, opts)
+
+  // if the decision is supposed to blur media,
+  // - have it apply to the view if it's on the account
+  // - otherwise ignore it
+  if (account.blurMedia) {
+    account.blur = true
+  }
+
+  // dont give profile.filter because that is meaningless
+  profile.filter = false
+
+  // downgrade based on authorship
+  if (!isModerationDecisionNoop(account) && account.did === opts.userDid) {
+    downgradeDecision(account, { alert: true })
+  }
+  if (!isModerationDecisionNoop(profile) && profile.did === opts.userDid) {
+    downgradeDecision(profile, { alert: true })
+  }
+
+  // derive avatar blurring from account & profile, but override for mutes because that shouldnt blur
+  let avatarBlur = false
+  let avatarNoOverride = false
+  if ((account.blur || account.blurMedia) && account.cause?.type !== 'muted') {
+    avatarBlur = true
+    avatarNoOverride = account.noOverride || profile.noOverride
+  } else if (profile.blur || profile.blurMedia) {
+    avatarBlur = true
+    avatarNoOverride = account.noOverride || profile.noOverride
+  }
+
+  // dont blur the account for blocking & muting
+  if (
+    account.cause?.type === 'blocking' ||
+    account.cause?.type === 'blocked-by' ||
+    account.cause?.type === 'muted'
+  ) {
+    account.blur = false
+    account.noOverride = false
+  }
+
+  return {
+    decisions: { account, profile },
+
+    // moderate all content based on account
+    account:
+      account.filter || account.blur || account.alert
+        ? toModerationUI(account)
+        : {},
+
+    // moderate the profile details based on the profile
+    profile:
+      profile.filter || profile.blur || profile.alert
+        ? toModerationUI(profile)
+        : {},
+
+    // blur or alert the avatar based on the account and profile decisions
+    avatar: {
+      blur: avatarBlur,
+      alert: account.alert || profile.alert,
+      noOverride: avatarNoOverride,
+    },
+  }
+}
+
+// posts
+// =
+
+export interface PostModeration {
+  decisions: {
+    post: ModerationDecision
+    account: ModerationDecision
+    profile: ModerationDecision
+    quote?: ModerationDecision
+    quotedAccount?: ModerationDecision
+  }
+  content: ModerationUI
+  avatar: ModerationUI
+  embed: ModerationUI
+}
+
+export function moderatePost(
```

Comment:
ah just thought of something here: do we ever want to make moderation decisions based on "repostedBy"?

I believe we strip out reposts from blocked/muted accounts, but what about reposts from a labeled account?

Reply:
that is a good question. to date we have not taken any moderation actions on reposts (and also not on like or follow records).

I think that yeah, we *should* probably consider both the re-posting *account* as well labels on the account and post record of the content itself. I'm not sure how common this is... reposts show up mostly in author feeds and following feeds (I think), so only if you follow or investigate the account. The user story would be something like, you followed somebody a long time ago, they have been labeled in a way your client would be filtering their posts, they re-post other similarly unwanted content, maybe bulk-reposting, that content ends up in your following feed.